### ISSUE TITLE
Phylogenetic backbones of sc2ts ARG and Viridian UShER tree

### ISSUE THREAD

**Issue by szhan** [jeromekelleher/sc2ts-paper#222]
Probably a simple approach to visually compare the backbones of the Viridian UShER tree and our pandemic-scale ARG is to leverage the Pango lineage roots, excluding the Pango recombinants. The Pango lineage roots are already labelled in the UShER tree, but it is trickier to  get the corresponding nodes in our ARGs. Suppose we do have the nodes identified, then we could simplify down to only those nodes (n = 2,131). For a cleaner view, we could exclude the less evolutionarily/epidemiologically relevant Pango lineages.

**Comment by szhan** [comment]
This can similarly be done for the global, all-time Nextstrain tree. Instead of using the Pango lineage labels, we would use the Nextstrain clade definitions. I think the easiest way may be to use the nucleotide definitions for each Nextstrain clade in order to identify the corresponding node in our ARGs, at least for the first pass. We should see the same clade hierarchy.

Some useful files for this analysis are here:
https://github.com/nextstrain/ncov/blob/master/defaults/clade_hierarchy.tsv
https://github.com/nextstrain/ncov/blob/master/defaults/clades_who.tsv
https://github.com/nextstrain/ncov/blob/master/defaults/clades.tsv

**Comment by szhan** [comment]
Also just encountered this list of mutations in the founder sequences of the Nextstrain clades assembled by Richard Neher. Tagging the list here in case they come in handy.

https://raw.githubusercontent.com/neherlab/SC2_variant_rates/cd6e016a511098123b6ce9ed874f58a7b789b34c/data/clade_gts.json

**Comment by hyanwong** [comment]
I did that in a simplistic way, by taking the MRCA of all the samples labelled as PangoNNN as the origin of that lineage, and collapsing those nodes. This is highly sensitive to errors in lineage designation, but easy to do. It does lead to a tree in which many Pango lineages share the same origin node, and it shows many large polytomies retained:

```python
import collections
tree = ts.at(21563)  # start of spike
pango_mrcas = {}
node_labels = collections.defaultdict(list)
for p, samples in ti.pango_lineage_samples.items():
    if not p.startswith("X") and not p == 'unknown':
        if len(samples) == 1:
            pango_mrcas[p] = samples[0]
        else:
            pango_mrcas[p] = tree.mrca(*samples)
        node_labels[pango_mrcas[p]].append(p)
node_labels = {k: "/".join([p for p in v]) for k, v in node_labels.items()}
sts = ts.simplify(list(set(pango_mrcas.values())), filter_nodes=False, keep_unary=True)
print("ARG simplified to pango non-X lineages has", sts.num_trees, "trees")
sts.at(21563).draw_svg(
    size=(2500, 1000),
    node_labels=node_labels,
    style=".leaf > .lab {text-anchor: start; transform: rotate(90deg) translate(6px)} .node text {font-size: 9px}",
    omit_sites=True,
)
```

<img width="1324" alt="Screenshot 2024-10-11 at 14 11 38" src="https://github.com/user-attachments/assets/60968aa8-ce51-446d-baa6-cc1289f849c3">

**Comment by hyanwong** [comment]
Jerome and I thought of another way, or intermediate hackiness: we could use Ana's lineage imputation method, and simply find the earliest node for each imputed pango lineage. We could test by looking both at the proportion of times that the standard lineage-defining mutations occur above this node (NB: if it is a unary node, we should include nodes below it too).

**Comment by szhan** [comment]
I wonder if we could just pick a representative sample for each scorpio label instead of the Pango labels. There are 33 [scorpio definitions](https://github.com/cov-lineages/constellations/tree/main/constellations/definitions), so quite manageable.

Here is the tanglegram used for the tskit meeting presentation comparing our sc2ts ARG and the Viridian UShER tree (to be remade). 28 of the 33 scorpio definitions are represented here. The scorpio definitions seem enough to capture the key relationships between the major variants, I think?

<img width="1016" alt="Image" src="https://github.com/user-attachments/assets/72414921-082a-4450-9dd6-c179a4209175" />

**Comment by jeromekelleher** [comment]
The Scorpio clades are the most important all right, I agree drawing trees capturing their relationships is the way to go for a main-text figure.

**Comment by szhan** [comment]
Some notes about the scorpio definitions.

# These differ from the main lineages by a single AA change
cA.23.1+E484K
cB.1.1.7+E484K
cB.1.617.2+K417N

# The parents of XBB and XE have their own definitions, for some reason
cXBB-parent1
cXBB-parent2
cXE-parent1
cXE-parent2

**Comment by szhan** [comment]
The above tanglegram just shows a sample with the lowest node ID (roughly the earliest sample) for each scorpio label. It seems to work okay, I think. Should we do something more complicated, like using imputed Pango labels of some ancestor as described above?

**Comment by hyanwong** [comment]
I have code to define the "originating node" for most Pango lineages. We could just use that?

**Comment by jeromekelleher** [comment]
This would be ideal if we can do the same thing for the Usher and sc2ts ARGs.

**Comment by szhan** [comment]
If we want to figure out the "origination nodes", we need to impute the scorpio labels instead, right?

**Comment by jeromekelleher** [comment]
You'd hope these are the same thing for such large-scale changes - the originating node for B.1.1.7 is surely the same as the originating node for Alpha scorpio

**Comment by hyanwong** [comment]
> You'd hope these are the same thing for such large-scale changes - the originating node for B.1.1.7 is surely the same as the originating node for Alpha scorpio

Exactly.

I think we can hand-craft the very few Pango lineage that are recombination nodes too.

**Comment by szhan** [comment]
Okay, so just discussing with Yan about the scorpio labels. It's not as clean as hoped, because scorpio labels are not really supergroups of Pango labels. It's some n-to-m mapping mess.

**Comment by jeromekelleher** [comment]
Yeesh, OK. Best stick with major pango clades then?

**Comment by szhan** [comment]
Yes, we were just discussing maybe doing a cutoff of the top X Pango labels that we can visualise in a tanglegram.

**Comment by szhan** [comment]
Putting this URL here in case it comes in handy. It has sets of curated lineage-defining mutations per VoC.
https://github.com/ukhsa-collaboration/variant_definitions/tree/main

**Comment by hyanwong** [comment]
Just returning to this. Here's the code (a bit slow to run, as I repeatedly simplify the ARG down to the respective pango samples, but it does the trick:

```python
ts = nb_utils.load()
ti = sc2ts.TreeInfo(ts)

import numpy as np

def representative_sample_node(simp_ts, pango):
    nodes_time = simp_ts.nodes_time
    valid = {}
    if simp_ts.num_samples == 0:
        return None
    # Find valid nodes: have same Pango type and > 50% descendant samples
    for tree in simp_ts.trees():
        for u in tree.nodes():
            tns = tree.num_samples(u) 
            if tns > simp_ts.num_samples/2 and simp_ts.node(u).metadata["Imputed_Viridian_pangolin"] == pango:
                if u not in valid:
                    valid[u] = (nodes_time[u], tns)
                else:
                    valid[u] = (nodes_time[u], max(tns, valid[u][1]))

    # Check through valid nodes for the best origination node
    if len(valid) == 0:
        print(pango, "has no valid origination nodes")
    else:
        orig_node = min(valid, key=lambda x: -valid[x][0])
        if valid[orig_node][1] / simp_ts.num_samples < 0.95:
            # Don't return a representative_sample_node if the origination node
            # accounts for < 95% of this pango samples
            print(
                f"{pango} with {len(valid)} valid origination nodes", 
                f"has oldest node {orig_node} rejected because it covers only",
                f"{valid[orig_node][1] / simp_ts.num_samples * 100:.2f}% of {pango} samples")
            return None
        else:
            # Find the oldest sample descendant of the origination node
            if simp_ts.node(orig_node).is_sample():
                return orig_node
            else:
                oldest_sample = max(
                    [
                        max(tree.samples(orig_node), key=lambda x: nodes_time[x])
                        for tree in simp_ts.trees(sample_lists=True)
                    ],
                    key=lambda x: nodes_time[x]
                )
                assert simp_ts.node(oldest_sample).metadata["Viridian_pangolin"] == pango
                return oldest_sample

# For speed, simplify the ARG down to a set of subgroups of pango samples
# and then further simplify the subgroup_ts to leave an ARG with just
# specific Pango samples
n_subgroup = int(np.sqrt(len(ti.pango_lineage_samples)))
batch = []
pango_representative_sample = {}
for i, k in tqdm(enumerate(ti.pango_lineage_samples), total=len(ti.pango_lineage_samples)):
    batch.append(k)
    if len(batch) == n_subgroup or i == len(ti.pango_lineage_samples) - 1:
        subgroup_ts = ts.simplify([u for b in batch for u in ti.pango_lineage_samples[b]], keep_unary=True, filter_nodes=False)
        for pango in batch:
            samples = ti.pango_lineage_samples[pango]
            simp_ts = subgroup_ts.simplify(samples, keep_unary=True, filter_nodes=False)
            best_nd = representative_sample_node(simp_ts, pango)
            if best_nd is not None:
                pango_representative_sample[pango] = best_nd
        batch = []

print(f"Found {len(pango_representative_sample)} pango sample nodes / {len{ti.pango_lineage_samples} pango groups")

sts = ts.simplify([v for k, v in best.items() if not k.startswith("X")], keep_unary=True, filter_nodes=False)
print("Simplified ARG to representative samples has {sts.num_trees} trees and {sts.num_samples} samples")
```

**Comment by hyanwong** [comment]
Closing, as now done with tanglegrams etc.
